simulator.py

```diff
@@ -0,0 +1,172 @@
+                 self.__pnl[index].trading_cost, self.__portfolio[index],
+                 self.__portfolio[index].position_np))
+        return result
+
+    def produce_portfolio(self, today, alpha_formula, data, price_dataset):
+        """
+        Find daily portfolio.
+        """
+        portfolio = alpha_formula(today, data)
+        portfolio.normalize()
+        portfolio.encode_np(price_dataset)
+        self.__portfolio.append(portfolio)
+
+    def produce_pnl(self, region, cost_function, target_data):
+        """
+        Calculate pnl for alphas.
+        This method will be called by simulation day by day while inputting
+        a new portfolio.
+
+        While calculating pnl,
+        we should take a look at our asset and trading cost.
+            (1) Total asset
+                We assume that our assets are the same everyday (equal to
+                self.initial_asset). To achieve this goal, we open a virtual
+                account to borrow/lend money to the investing account (this
+                account is self.cur_asset[self.base]).
+            (2) Trading cost
+                There are different functions about calculating trading cost
+                in different investing targets.
+                This method should import different modules while targets are
+                different.
+
+                In our template function, we calculate trading cost while
+                selling out the asset.
+                The trading_cost function could be built by alpha writers.
+                The API will adjust in the future, while the fuction needs
+                more data.
+        """
+        today = self.__portfolio[-1]
+        total_cost = 0
+        total_profit = 0
+        trading_value = 0
+        for instrument, proportion in today.items():
+            if target_data[instrument][-1].sell is not None:
+                now = self.quantity(
+                    proportion,
+                    proportion / target_data[instrument][-1].sell
+                )
+                trading_value += now.value
+                profit, cost = cost_function(
+                    self.cur_asset[instrument],
+                    now,
+                    target_data[instrument][-1]
+                )
+                self.cur_asset[instrument] = now
+                total_profit += profit
+                total_cost += cost
+        basic_value = 1 - trading_value
+        total_profit += self.cur_asset[region].value - basic_value
+        self.cur_asset[region] = self.quantity(basic_value, basic_value)
+        self.__pnl.append((today.date, total_profit - total_cost, total_cost))
+
+
+class Simulator(SimulateMethod):
+    """
+    Simulate the alpha.
+    """
+    def __init__(self, settings, data):
+        SimulateMethod.__init__(
+            self,
+            settings.target[1],
+            list(data['target_data'].keys()))
+        self.alpha_formula = None
+        self.data = None
```

Comment:
Instead of having three different variables (data, target_data, workdays), I would suggest to pack them in a dict, just like what we do in `simulation.extract_data`.